Add 'Download' button to record set attachment viewer

alesan99 opened the pull request:
Fixes #609 

![image](https://github.com/user-attachments/assets/acc1bf6f-1b0f-43e0-9f83-6c16787c09fc)

This PR adds two new features to speed up attachment downloading.
1. When viewing the attachment gallery for a record set, a 'download all' button is now available. Clicking it should start downloading a zip file containing all the attachments, or it will download the attachment directly if there is only one.
2. I also added new small download buttons to each individual attachment, which speeds up downloading a selection of attachments. This was applied universally.

Additional Notes:
The URL to download all the attachments is `attachment_gw/download_all/`. Couldn't come up with a better name, so if you have a better name let me know!
If you download multiple attachments with the same name they will be renamed to OriginalName_AssetServerName.EXT.  I can also change this if it doesn't seem good enough.

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good and
      self-explanatory (or properly documented)
- [ ] Add automated tests
- [x] Add relevant issue to release milestone
- [x] Add relevant documentation (Tester - Dev) 

### Testing instructions

<!-- What are the steps to verify the fixes/changes in this PR? -->
<!-- Can part of that be replaced by automated tests? -->

Test 'Download All' button:
- Use a database that has a lot of attachments
- In QB, run a Query on Collection Objects that have attachments
- Browse in Forms
- Click the attachment gallery icon on the header
- Click the 'Download All' button
- [ ] Download is successful
- [ ] Verify that the zip file contains all the requested attachments with the correct filenames.
- [ ] Make sure the download all button is only enabled when there are attachments to download.
- [ ] With many attachments, make sure the download all button doesn't break when not all attachments have been loaded yet.
- Try out the button in different attachments galleries, like in individual records or in record sets.
- [ ] Try to download multiple attachments with the same file name. Make sure they all get downloaded correctly. They should be named `fileName_2.EXT` and so on.
Test small download buttons:
- [ ] Verify that the download button works in different contexts (Attachment viewer, Record set attachment gallery, etc.)

grantfitzsimmons commented:
🐐

grantfitzsimmons commented on specifyweb/attachment_gw/views.py:
```diff
@@ -296,6 +305,69 @@ def proxy(request):
+            }))
+
+    # # Send zip as a notification? It may be possible to automatically send it as a download? Maybe not a good idea if server is under load.
+    # # Or stream the zip file back to the client with StreamingHttpResponse
```
>These archives could potentially be huge so it might not a good idea to keep them around in DEPOSITORY_DIR.

There needs to be a mechanism to ensure these files do not remain on the server indefinitely. In theory, we could have the download use the same name and continually overwrite it in its simplest implementation...

I think streaming it to the client is the way to go if possible. Otherwise, we would need to make sure we're cleaning up regularly.

alesan99 replied:
Yeah storing the file for the notification was seeming not ideal. After considering my options a bit I decided to stream the file back to the client.
On top of that it feels a lot better now too! It save a couple clicks.

alesan99 commented:
Looking for feedback on these👇! This PR ended up being a bit bigger than I thought so I want to make sure everything is right, 

Dev Note:
To my dismay, it looks like receiving zip files with a POST request wasn't possible with the existing utilities :(. The zip file would get turned into a string at multiple points of the process.
I had to change the ajax fetch utility a bit (and downloadFile). Once a response is read with response.text(), it can no longer be read with response.blob(), so I couldn't extract the original binary data further in the process.
I prioritized minimizing code impact mainly, but the code still might not be ideal.
I'm not comfortable messing with such important code, so I am looking on feedback on those changes especially.

Additional Notes:
The URL to download all the attachments is `attachment_gw/download_all/`. Couldn't come up with a better name, so if you have a better name let me know!
If you download multiple attachments with the same name they will be renamed to OriginalName_AssetServerName.EXT.  I can also change this if it doesn't seem good enough.

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Attachments/RecordSetAttachment.tsx:
```diff
@@ -81,6 +85,36 @@ export function RecordSetAttachments<SCHEMA extends AnySchema>({
 
+  const handleDownloadAllAttachments = async (): Promise<void> => {
+    if (attachmentsRef.current === undefined) return;
+    const attachmentLocations: readonly string[] = attachmentsRef.current.attachments
```
don't need explicit type annotation - typescript can infer it

```suggestion
    const attachmentLocations = attachmentsRef.current.attachments
```

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Attachments/RecordSetAttachment.tsx:
```diff
@@ -81,6 +85,36 @@ export function RecordSetAttachments<SCHEMA extends AnySchema>({
+    const attachmentLocations: readonly string[] = attachmentsRef.current.attachments
+      .map((attachment) => attachment.attachmentLocation)
+      .filter((name): name is string => name !== null);
+    const origFilenames: readonly string[] = attachmentsRef.current.attachments
```
same

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Attachments/RecordSetAttachment.tsx:
```diff
@@ -81,6 +85,36 @@ export function RecordSetAttachments<SCHEMA extends AnySchema>({
+      });
+
+      if (response.status === Http.OK) {
+        downloadFile('attachments.zip', response.data);
```
(optional) could we pick a better name? maybe use record set name?

alesan99 replied:
Went with "Attachments - RecordSetName.zip" 👍

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Attachments/RecordSetAttachment.tsx:
```diff
@@ -110,7 +144,17 @@ export function RecordSetAttachments<SCHEMA extends AnySchema>({
-            <Button.DialogClose>{commonText.close()}</Button.DialogClose>
+            <>
+              <Button.Info
+                disabled={fetchedCount.current !== records.length || records.length <= 1}
```
it may not be clear to uset that all attachments need to be bulked fetched by front-end for them to be downloadable
could you add a `title={}` with explanation to this button when it is disabled?

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Molecules/FilePicker.tsx:
```diff
@@ -137,26 +137,34 @@ export function FilePicker({
+        );
+        element.setAttribute('download', fileName);
+      } else {
+        url = window.URL.createObjectURL(data);
```
nice! didn't know this was a thing

```suggestion
        url = URL.createObjectURL(data);
```

maxpatiiuk commented on specifyweb/frontend/js_src/lib/components/Molecules/FilePicker.tsx:
```diff
@@ -137,26 +137,34 @@ export function FilePicker({
     iframe.classList.add('absolute', 'hidden');
     iframe.addEventListener('load', () => {
       if (iframe.contentWindow === null || fileDownloaded) return;
+      let url;
```
add `:string|undefined;` type annotation?

maxpatiiuk commented on specifyweb/frontend/js_src/lib/utils/ajax/index.ts:
```diff
@@ -135,16 +137,21 @@ export async function ajax<RESPONSE_TYPE = string>(
   })
-    .then(async (response) => Promise.all([response, response.text()]))
+    .then(async (response) => 
+       (acceptBlobResponse ?
+        Promise.all([response, response.blob()]) : 
+        Promise.all([response, response.text()])
```
```suggestion
       Promise.all([response, acceptBlobResponse ? response.blob() : response.text() ])
```

maxpatiiuk reviewed (approved):
great job on a complex feature!

the changes in ajax() look good - this is exactly how I would have done it 🤗

sharadsw reviewed (approved):
Looks good to me. Good job!

pashiav reviewed (changes requested):
- [X] Download is successful
- [X] Verify that the zip file contains all the requested attachments with the correct filenames.
- [ ] Make sure the download all button is only enabled when there are multiple attachments to download.
- [X] With many attachments, make sure the download all button doesn't break when not all attachments have been loaded yet.
- [X] Try to download multiple attachments with the same file name. Make sure they all get downloaded correctly. They should be named `OriginalFileName_InternalFileName.EXT`.
Test small download buttons:
- [X] Verify that the download button works in different contexts (Attachment viewer, Record set attachment gallery, etc.)
---

Great work so far!


For: 

> - [ ] Make sure the download all button is only enabled when there are multiple attachments to download.

The **Download All** button is a little confusing because it's disabled for a singular attachment/record. It says it downloads **all** attachments when you hover over it, so I think it should include and work for singular attachments too. Removing the button when there is only one attachment could work as well. @specify/ux-testing what do you think?

<img width="668" alt="Screenshot 2025-01-22 at 8 43 03 AM" src="https://github.com/user-attachments/assets/f20fd023-c4ab-47ff-996a-52cba646c7f9" />

<br><br>

Also I found that if you have 1 record with 1 attachment, but select multiple records which do not have any attachments, the **Download All** button is functional and will create a zip file of only 1 attachment.

**To reproduce:**
1. Select a record with 1 attachment in QB
2. Select other records with no attachments in addition to it
3. Click Browse in Forms -> Attachment Gallery icon -> Download All
4. See a zip file has been downloaded with 1 attachment


https://github.com/user-attachments/assets/09c67385-af3b-4d59-a283-e71b24c34ff7

Areyes42 reviewed (changes requested):
- [ ]  Download is successful
-  [x]  Verify that the zip file contains all the requested attachments with the correct filenames.
-  [ ]  Make sure the download all button is only enabled when there are multiple attachments to download.
-  [x]  With many attachments, make sure the download all button doesn't break when not all attachments have been loaded yet.
-  [x] Try out the button in different attachments galleries, like in individual records or in record sets.
-   [x] Try to download multiple attachments with the same file name. Make sure they all get downloaded correctly. They should be named `fileName_2.EXT` and so on.  
    Test small download buttons:
-  [x]  Verify that the download button works in different contexts (Attachment viewer, Record set attachment gallery, etc.)
***
Great work so far, Alejandro!

For https://github.com/specify/specify7/pull/6052#pullrequestreview-2567895305, I agree that keeping the **Download All** option for records with only one attachment is more consistent, even if it feels redundant. I also noticed that when browsing a single form with multiple attachments, the **Download All** button is also disabled.

![Screenshot 2025-01-22 at 1 42 05 PM](https://github.com/user-attachments/assets/c79cdd80-982b-4395-8715-827c0d01134a)


Additionally, I kept running into a crash dialog while using the **Download All** button on the `KUFish` databases. Initially, I thought it might be related to the asset server configuration, preventing it from properly fetching the assets. However, I was able to download each asset individually without issues, so I'm thinking it might be another issue. The errors only occurred when trying to download all attachments at once.

https://github.com/user-attachments/assets/21186c1e-50f4-4b0e-a1b8-7ef3c6a1fcd0

[Specify 7 Crash Report - 2025-01-22T20_11_48.389Z.txt](https://github.com/user-attachments/files/18511360/Specify.7.Crash.Report.-.2025-01-22T20_11_48.389Z.txt)

alesan99 commented:
Ok I addressed everything mentioned and its ready for a retest👌
I changed the logic for when the button is enabled, so it should now be enabled as long as 1 attachment has been loaded.

> The **Download All** button is a little confusing because it's disabled for a singular attachment/record. It says it downloads **all** attachments when you hover over it, so I think it should include and work for singular attachments too. Removing the button when there is only one attachment could work as well. @specify/ux-testing what do you think?

I changed it so you can download a single attachment as well. Let me know if it works as expected!

> Additionally, I kept running into a crash dialog while using the Download All button on the KUFish databases. Initially, I thought it might be related to the asset server configuration, preventing it from properly fetching the assets. However, I was able to download each asset individually without issues, so I'm thinking it might be another issue. The errors only occurred when trying to download all attachments at once.

I'm not sure what is happening here exactly, I tried to recreate it locally but I couldn't. Either way a completely blank error isn't good so I tried to add a better error if something went wrong.

KUFish looks like it might have something weird going on with the filenames so I added some handling for that, lmk if its fixed now!

Areyes42 reviewed (changes requested):
Download all button for single attachments looks good and works properly, yay!

For my previous issue with KUFish, I'm still able to recreate it, but I think I've narrowed it down to this specific [collection object attachment](https://kufish20240606-issue-609.test.specifysystems.org/specify/view/collectionobject/38531/). This looks to be a one-off scenario, and I don't see any obvious issue with the filename or extension. Not sure if it's an error with this specific db or with the backend fetching, but all the other db's I've tested have not had this error.

Also, I didn't realize this before, but the download all button is not in the form attachments viewer—could it be added for consistency?

https://github.com/user-attachments/assets/3593395d-2a0e-4176-be5a-9412aad0f9a9

pashiav reviewed (approved):
- [X] Download is successful
- [X] Verify that the zip file contains all the requested attachments with the correct filenames.
- [X] Make sure the download all button is only enabled when there are attachments to download.
- [X] With many attachments, make sure the download all button doesn't break when not all attachments have been loaded yet.
- [X] Try to download multiple attachments with the same file name. Make sure they all get downloaded correctly. They should be named `fileName_2.EXT` and so on.
- [X] Verify that the download button works in different contexts (Attachment viewer, Record set attachment gallery, etc.)
---

Great work!

---

I noticed this weird behavior that is not caused by this PR but thought I'd leave it as a note for other testers:

My workflow:
- add one attachment each to 3 new COs
-  Attachment viewer -> Download All --- works as expected
- change all 3 attachments to be named “1” --- notice that the Save button isn't enabled after the changes
- Attachment Viewer-> Download All -> open zip and see files like this:
<img width="644" alt="Screenshot 2025-02-05 at 1 07 10 PM" src="https://github.com/user-attachments/assets/f6963ddb-75be-4440-b832-97e03bcb1511" />

After further testing I found that the Save button not enabling is an issue and regression in production. I'll update this comment with the issue after writing it up.

alesan99 commented:
> For my previous issue with KUFish, I'm still able to recreate it, but I think I've narrowed it down to this specific [collection object attachment](https://kufish20240606-issue-609.test.specifysystems.org/specify/view/collectionobject/38531/). This looks to be a one-off scenario, and I don't see any obvious issue with the filename or extension. Not sure if it's an error with this specific db or with the backend fetching, but all the other db's I've tested have not had this error.

Found the issue, looks like one of the attachments was being downloaded with the entire file path as its name which isn't good.
I fixed how the file name is retrieved and it looks like it works now! And the front end should be able to receive errors now so its easier to debug in the future as well

I also added the download all button to the other form attachments viewer 👍👍👍

emenslin reviewed (approved):
- [x] Download is successful
- [x] Verify that the zip file contains all the requested attachments with the correct filenames.
- [x] Make sure the download all button is only enabled when there are attachments to download.
- [x] With many attachments, make sure the download all button doesn't break when not all attachments have been loaded yet.
- [x] Try to download multiple attachments with the same file name. Make sure they all get downloaded correctly. They should be named `fileName_2.EXT` and so on.
Test small download buttons:
- [x] Verify that the download button works in different contexts (Attachment viewer, Record set attachment gallery, etc.)

Looks good!

Areyes42 reviewed (approved):
-  [x] Download is successful
-   [x] Verify that the zip file contains all the requested attachments with the correct filenames.
-  [x]  Make sure the download all button is only enabled when there are attachments to download.
-   [x] With many attachments, make sure the download all button doesn't break when not all attachments have been loaded yet.
-  [x]  Try to download multiple attachments with the same file name. Make sure they all get downloaded correctly. They should be named `fileName_2.EXT` and so on.  
    Test small download buttons:
-  [x]  Verify that the download button works in different contexts (Attachment viewer, Record set attachment gallery, etc.)
***
Hey, great job on this PR!! All my previous comments got addressed and everything looks good! 👍

specifysoftware commented:
This pull request has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/specify-7-10-2-release-announcement/2455/1
